storage/transfermanager/downloader.go

```diff
@@ -784,3 +894,57 @@ func setUsageMetricHeader(ctx context.Context, method string) context.Context {
 	header := fmt.Sprintf("%s/%s", usageMetricKey, method)
 	return callctx.SetHeaders(ctx, xGoogHeaderKey, header)
 }
+
+type crc32cPiece struct {
+	sum    uint32 // crc32c checksum of the piece
+	length int64  // number of bytes in this piece
+}
+
+// joinCRC32C pieces together the initial checksum with the orderedChecksums
+// provided to calculate the checksum of the whole.
+func joinCRC32C(initialChecksum uint32, orderedChecksums []crc32cPiece) uint32 {
+	base := initialChecksum
+
+	zeroes := make([]byte, maxChecksumZeroArraySize)
+	for _, part := range orderedChecksums {
+		// Precondition Base (flip every bit)
+		base ^= 0xFFFFFFFF
+
+		// Zero pad base crc32c. To conserve memory, do so with only maxChecksumZeroArraySize
+		// at a time. Reuse the zeroes array where possible.
+		var padded int64 = 0
+		for padded < part.length {
+			desiredZeroes := min(part.length-padded, maxChecksumZeroArraySize)
+			base = crc32.Update(base, crc32.MakeTable(crc32.Castagnoli), zeroes[:desiredZeroes])
+			padded += desiredZeroes
+		}
+
+		// Postcondition Base (same as precondition, this switches the bits back)
+		base ^= 0xFFFFFFFF
+
+		// Bitwise OR between Base and Part to produce a new Base
+		base ^= part.sum
+	}
+	return base
+}
+
+func fullObjectRead(r *DownloadRange) bool {
+	return r == nil || (r.Offset == 0 && r.Length < 0)
+}
+
+func checksumObject(ctx context.Context, object *storage.ObjectHandle, got, want uint32) error {
+	// If we don't have a CRC32C from the reader, do a metadata call to get it.
```

Comment:
Do we still need this? If so under what circumstances? Also, we should explicitly note that generation must be set on `object` for this to work.

Reply:
Hmm, I guess this is just a fallback, but yeah I don't think it actually gets called - but then again I haven't done super extensive testing.

Reply:
I think since Reader doesn't have a fallback if the header is unavailable, we shouldn't have one here either.